packages/algorand-js/src/runtime/runtime.ts

```diff
@@ -0,0 +1,215 @@
+   * app's global data for application id: appId, throw error otherwise
+   * @param appId: current application id
+   * @param key: key to fetch value of from local state
+   * @param value: key to fetch value of from local state
+   */
+  updateGlobalState (appId: number, key: Uint8Array, value: StackElem): SSCStateSchema[] {
+    const appDelta = this.assertAppDefined(appId);
+    const globalState = appDelta["global-state"];
+
+    const data = getKeyValPair(key, value); // key value pair to put
+    const idx = globalState.findIndex(schema => compareArray(schema.key, key));
+    if (idx === -1) {
+      globalState.push(data); // push new pair if key not found
+    } else {
+      globalState[idx].value = data.value; // update value if key found
+    }
+    appDelta["global-state"] = globalState; // save updated state
+
+    assertValidSchema(appDelta["global-state"], appDelta["global-state-schema"]); // verify if updated schema is valid by config
+    return globalState;
+  }
+
+  /**
+   * Description: set accounts for context as {address: SdkAccount}
+   * @param accounts: array of account info's
+   */
+  createStatefulContext (accounts: SdkAccount[]): void {
+    for (const acc of accounts) {
+      this.store.state.accounts.set(acc.address, acc);
+
+      for (const app of acc.createdApps) {
+        this.store.state.globalApps.set(app.id, app.params);
+      }
+    }
+  }
+
+  /**
+   * Description: creates a new transaction object from given execParams
+   * @param txnParams : Transaction parameters for current txn or txn Group
+   */
+  createTxnContext (txnParams: ExecParams | ExecParams[]): void {
+    // if txnParams is array, then user is requesting for a group txn
+    if (Array.isArray(txnParams)) {
+      if (txnParams.length > 16) {
+        throw new Error("Maximum size of an atomic transfer group is 16");
+      }
+
+      const txns = [];
+      for (const txnParam of txnParams) { // create encoded_obj for each txn in group
+        const mockParams = mockSuggestedParams(txnParam.payFlags);
+        const tx = mkTransaction(txnParam, mockParams);
+
+        // convert to encoded obj for compatibility
+        const encodedTxnObj = tx.get_obj_for_encoding() as Txn;
+        encodedTxnObj.txID = tx.txID();
+        txns.push(encodedTxnObj);
+      }
+      assignGroupID(txns); // assign unique groupID to all transactions in the array/group
+      this.store.gtxs = txns;
+      this.store.tx = txns[0]; // by default current txn is the first txn
+    } else {
+      // if not array, then create a single transaction
+      const mockParams = mockSuggestedParams(txnParams.payFlags);
+      const tx = mkTransaction(txnParams, mockParams);
+
+      const encodedTxnObj = tx.get_obj_for_encoding() as Txn;
+      encodedTxnObj.txID = tx.txID();
+      this.store.tx = encodedTxnObj; // assign current txn
+      this.store.gtxs = [this.store.tx]; // assign single txn to grp
+    }
+  }
```

Comment:
This has to be changed. When I was talking about context, I was specifically mention that we need to protect the state and separate it from transactions.
Context should have:
* copy of store (so when transaction will fail, we will be able to remove context without modifying the store). We need to protect the store. If transaction fails, it MUST NOT modify the store - hence we copy it to the context.
* transactions and all other required data - please don't put transactions in store.

Reply:
Ok, i am separating `[tx, gtxs]` from store.
But i don't understand why. Transaction data is READ ONLY (opcodes only get a txn property and put to stack. eg. `txn Fee`).

Reply:
We don't put transactions is state. If we need to keep the transaction data for whatever reason (but not in store) then it's better to copy it. Otherwise, we don't need to copy and we can pass it to the context.

Reply:
Ok, thanks